pytorch_widedeep/losses.py

```diff
@@ -7,6 +7,131 @@
+            2 - p
+        )
+        return torch.mean(loss)
+
+
+class QuantileLoss(nn.Module):
+    """
+    Quantile loss, i.e. a quantile of ``q=0.5`` will give half of the mean absolute error as it is calcualted as
+
+    Defined as ``max(q * (y-y_pred), (1-q) * (y_pred-y))``
+    All credits go to `pytorch-forecasting
+    <https://pytorch-forecasting.readthedocs.io/en/latest/_modules/pytorch_forecasting/metrics.html#QuantileLoss>`
+    """
+
+    def __init__(
+        self,
+        quantiles: List[float] = [0.02, 0.1, 0.25, 0.5, 0.75, 0.9, 0.98],
+    ):
+        """
+        Quantile loss
+
+        Args:
+            quantiles: quantiles for metric
+        """
+        super().__init__()
+        self.quantiles = quantiles
+
+    def forward(self, input: Tensor, target: Tensor) -> Tensor:
+        assert input.shape == torch.Size([target.shape[0], len(self.quantiles)]), (
+            f"Wrong shape of input, pred_dim of the model that is using QuantileLoss must be equal "
+            f"to number of quantiles, i.e. {len(self.quantiles)}."
+        )
+        target = target.view(-1, 1).float()
+        losses = []
+        for i, q in enumerate(self.quantiles):
+            errors = target - input[..., i]
+            losses.append(torch.max((q - 1) * errors, q * errors).unsqueeze(-1))
+        loss = torch.cat(losses, dim=2)
+
+        return torch.mean(loss)
+
+
+class ZILNLoss(nn.Module):
+    r"""Adjusted implementation of the `Zero Inflated LogNormal loss
+    <https://arxiv.org/pdf/1912.07753.pdf>` and its `code
+    <https://github.com/google/lifetime_value/blob/master/lifetime_value/zero_inflated_lognormal.py>`
+    """
+
+    def __init__(self):
+        super().__init__()
+
+    def forward(self, input: Tensor, target: Tensor) -> Tensor:
+        r"""
+        Parameters
+        ----------
+        input: Tensor
+            input tensor with predictions (not probabilities) with spape (N,3), where N is the batch size
+        target: Tensor
+            target tensor with the actual classes
+
+        Examples
+        --------
+        >>> import torch
+        >>>
+        >>> from pytorch_widedeep.losses import ZILNLoss
+        >>>
+        >>> # REGRESSION
+        >>> target = torch.tensor([[0., 1.5]]).view(-1, 1)
+        >>> input = torch.tensor([[.1, .2, .3], [.4, .5, .6]])
+        >>> ZILNLoss()(input, target)
+        tensor(1.3114)
+        """
+        positive = target > 0
+        positive = positive.float()
+
+        assert input.shape == torch.Size(
+            [target.shape[0], 3]
+        ), "Wrong shape of input, pred_dim of the model that is using ZILNLoss must be equal to 3."
+        positive_input = input[..., :1]
+
+        classification_loss = F.binary_cross_entropy_with_logits(
+            positive_input, positive, reduction="none"
+        ).flatten()
+
+        loc = input[..., 1:2]
+
+        # when using max the two input tensors (input and other) have to be of
```

Comment:
Added these two lines to make sure we can use `torch.max(input, other)` and therefore is compatible with older versions of pytorch